### PR TITLE
AP_AHRS: UAVCAN Position and attitude broadcast v3

### PR DESCRIPTION
**Pull request by EShamaev**
With changes and updates from https://github.com/ArduPilot/ardupilot/pull/7827 and https://github.com/ArduPilot/ardupilot/pull/6528

### CODE REVIEW

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 1]
```diff
@@ -71,9 +73,231 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
...
+        _fix_out_sem->give();
+    }
+    
+    if ((sem_ret = _att_out_sem->take(1)) && (attitude_out_array[_uavcan_i] != nullptr)) {
+        uavcan::Timestamp ts;
+        ts.usec = AP_HAL::micros64();
+        _att_state.timestamp = ts;
+
+        Quaternion qt;
+        Matrix3f rot = ahrs.get_rotation_body_to_ned();
```
Comment (khancyr) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: const

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 2]
```diff
@@ -71,9 +73,231 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
...
+        qt.from_rotation_matrix(rot);
+        _att_state.orientation_xyzw[0] = qt.q1;
+        _att_state.orientation_xyzw[1] = qt.q2;
+        _att_state.orientation_xyzw[2] = qt.q3;
+        _att_state.orientation_xyzw[3] = qt.q4;
+
+        // TODO: extract from EKF
+        //_att_state.orientation_covariance
+
+        Vector3f av = ahrs.get_gyro();
```
Comment (khancyr) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: const

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 3]
```diff
@@ -71,9 +73,231 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
...
+        _att_state.angular_velocity[0] = av.x;
+        _att_state.angular_velocity[1] = av.y;
+        _att_state.angular_velocity[2] = av.z;
+
+        // TODO: extract from EKF
+        //_att_state.angular_velocity_covariance
+
+        Vector3f la = ahrs.get_accel_ef();
```
Comment (khancyr) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: const

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 4]
```diff
@@ -658,6 +929,8 @@ void AP_UAVCAN::do_cyclic(void)
 		return;
 	}
 
+    uint32_t now = AP_HAL::millis();
```
Comment (khancyr) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: const

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 5]
```diff
@@ -71,9 +71,231 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("ESC_BM", 3, AP_UAVCAN, _esc_bm, 255),
 
+    // @Param: FIX_R
+    // @DisplayName: UAVCAN GNSS Fix broadcast rate
+    // @Description: UAVCAN GNSS Fix broadcast rate in times per second. 0 - disable broadcast.
+    // @Range: 0 20
+    // @User: Advanced
+    AP_GROUPINFO("FIX_R", 4, AP_UAVCAN, _broadcast_fix_rate, 10),
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: I'm reluctant to have this broadcast at 10Hz by default. This PR puts 20-30Hz of packets onto the bus when in most cases no one wants that data

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 6]
```diff
@@ -71,9 +71,231 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
...
+static uavcan::Publisher<uavcan::equipment::ahrs::Solution> *attitude_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+
+static uavcan::equipment::gnss::Fix _fix_state;
+static uavcan::equipment::gnss::Fix2 _fix2_state;
+static uavcan::equipment::ahrs::Solution _att_state;
+
+void AP_UAVCAN::UAVCAN_AHRS_update(const AP_AHRS_NavEKF &ahrs)
+{
+    bool sem_ret;
+    if ((sem_ret = _fix_out_sem->take(1)) &&
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: I get a little lost in this long function. Would you consider changing it to something like:

fix_update()
fix2_update()
ahrs_update()

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 7]
```diff
@@ -687,6 +911,37 @@ void AP_UAVCAN::do_cyclic(void)
...
+    if (att_out_sem_take() &&
+            attitude_out_array[_uavcan_i] != nullptr &&
+            _broadcast_att_rate != 0 &&
+            (now - att_out_send_last_ms) >= att_out_send_delta_ms) {
+        att_out_send_last_ms = now;
+        att_out_send_delta_ms = 1000U / (uint16_t)_broadcast_att_rate;
+        attitude_out_array[_uavcan_i]->broadcast(_att_state);
+
+        att_out_sem_give();
+	}
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: same with do_cyclic. It was gettign really messy with the ESC send stuff so I refactored it so the flow was obvious. Woudl be nice to follow the above examples with a fix_out_send() and att_out_send()

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 5 above)
Comment (EShamaev) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Agree. It is used only by few MAPIR camera users.

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 6 above)
Comment (EShamaev) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Yes, that's long but no point in separating. It's totally sequential and separation will not change anything.

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 7 above)
Comment (EShamaev) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Same reasons. It's really straight forward.

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 8]
```diff
@@ -511,9 +736,26 @@ bool AP_UAVCAN::try_init(void)
                     rgb_led[_uavcan_i] = new uavcan::Publisher<uavcan::equipment::indication::LightsCommand>(*node);
                     rgb_led[_uavcan_i]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
                     rgb_led[_uavcan_i]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
-
                     _led_conf.devices_count = 0;
 
+                    if(_broadcast_bm & AP_UAVCAN_BROADCAST_POSITION_FIX) {
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: I think something was lost in the rebase. No param sets _broadcast_bm so these subscribers will never init. We should always init them and just control the rates in the update/cyclic. Keeping like this requires a reboot if you turn it on after the boot which is not desirable - It should just start working.

Comment (EShamaev) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: It was not lost in rebase, but a bug that is a left over after changing the rate logic.

**Comment by magicrub** [comment]
Seems like there's excess work happening here.
AP_AHRS_NavEKF::update() is calling need_AHRS_update() very fast, but that data is only ever used at the param interval. Why not fetch the data on demand in do_cyclic? There's a good chance we'll want them at different rates too, like ATT at >=50Hz and FIX at 5Hz (and FIX2 at 1Hz?)

**Comment by magicrub** [comment]
I'd like to add the opposite of this, the ability to receive and parse the ATT/FIX message

**Comment by magicrub** [comment]
Please consider adding [this commit](https://github.com/magicrub/ardupilot/commit/9603ebc02dc106db9d405513a910f77c97dfb9d0). I put it on my own branch instead of pushing to yours so we don't delay the PR any further. I'll test this as soon as my babel arrives

**Comment by magicrub** [comment]
I added a commit taking the assignment out of the conditional if statement.

**Comment by magicrub** [comment]
any movement happening here?

**Comment by Linjieqiang** [comment]
any movement happening here?

**Comment by Linjieqiang** [comment]
any movement happening here?

**Comment by tridge** [comment]
@EShamaev this looks like you were trying to create a CAN peripheral, before we had AP_Periph? Do you still want to pursue this or are you happy with AP_Periph?

**Comment by bugobliterator** [comment]
we have had no response on this PR for quite a while, should be closed? @tridge

**Comment by khancyr** [comment]
I will close this for now. Thanks for the effort, if somebody want to rework this to integrate it, feel free to open a new PR with the changes.

**Pull request closed without merging**